src/synpp/pipeline.py

```diff
@@ -662,122 +673,84 @@ def run(definitions, config = {}, working_directory = None, flowchart_path = Non
 
                 ephemeral_counts[hash] += 1
 
-    # 3) Load information about stages
-    meta = {}
-
-    if not working_directory is None:
-        try:
-            with open("%s/pipeline.json" % working_directory) as f:
-                meta = json.load(f)
-                logger.info("Found pipeline metadata in %s/pipeline.json" % working_directory)
-        except FileNotFoundError:
-            logger.info("Did not find pipeline metadata in %s/pipeline.json" % working_directory)
-
     # 4) Devalidate stages
     sorted_cached_hashes = sorted_hashes - ephemeral_counts.keys()
     stale_hashes = set()
 
+    # Get current validation tokens
+    current_validation_tokens = {
+        stage_id:
+        str(
+            registry[stage_id]["wrapper"].validate(
+                ValidateContext(registry[stage_id]["config"], get_cache_directory_path(working_directory, stage_id))
+            )
+        ) for stage_id in sorted_hashes
+    }
+
+    # Cache mapper between stage id and cache id.
+    cache_ids = {stage_id: get_cache_prefix(stage_id, source_digests[stage_id]) + "__" + str(current_validation_tokens[stage_id]) for stage_id in sorted_hashes}
```

Comment:
Why do we need this `"__" + validation_tokens`, can't this just be included in the hash? Like hashing the `get_cache_prefix(...)` and the validation tokens together?